plugins/woocommerce-blocks/assets/js/data/checkout/push-changes.ts

```diff
@@ -0,0 +1,110 @@
+/**
+ * External dependencies
+ */
+import { debounce } from '@woocommerce/base-utils';
+import { select, dispatch } from '@wordpress/data';
+import isShallowEqual from '@wordpress/is-shallow-equal';
+
+/**
+ * Internal dependencies
+ */
+import { STORE_KEY } from './constants';
+import { processErrorResponse } from '../utils';
+
+// This is used to track and cache the local state of push changes.
+const localState = {
+	// True when the checkout data has been initialized.
+	checkoutDataIsInitialized: false,
+	// True when a push is currently happening to avoid simultaneous pushes.
+	doingPush: false,
+	// Local cache of the last pushed checkoutData used for comparisons.
+	checkoutData: {
+		orderNotes: '',
+		additionalFields: {},
+	},
+};
+
+/**
+ * Initializes the checkout data cache on the first run.
+ */
+const initialize = () => {
+	const store = select( STORE_KEY );
+	localState.checkoutData = {
+		orderNotes: store.getOrderNotes(),
+		additionalFields: store.getAdditionalFields(),
+	};
+	localState.checkoutDataIsInitialized = true;
+};
+
+/**
+ * Function to dispatch an update to the server.
+ */
+const updateCheckoutData = (): void => {
+	if ( localState.doingPush ) {
+		return;
+	}
+
+	// Prevent multiple pushes from happening at the same time.
+	localState.doingPush = true;
+
+	// Get current checkout data from the store
+	const store = select( STORE_KEY );
+	const newCheckoutData = {
+		orderNotes: store.getOrderNotes(),
+		additionalFields: store.getAdditionalFields(),
+	};
+
+	// Do we need to push anything?
+	if ( isShallowEqual( localState.checkoutData, newCheckoutData ) ) {
+		localState.doingPush = false;
+		return;
+	}
+
+	// Update local cache
+	localState.checkoutData = newCheckoutData;
+
+	dispatch( STORE_KEY )
+		.updateDraftOrder( {
```

Comment:
Is checkout essentially blocked when this is taking place?

Reply:
You can still interact with the UI, it happens in the background async

Reply:
Should be the same as the cart fetch when we update the customer data